### PR TITLE
[WIP] ProMPs

### PR DESCRIPTION
**Pull request by AlexanderFabisch**

### CODE REVIEW

**Comment by AlexanderFabisch** [comment]
Please remove the logo.

**Comment by AlexanderFabisch** [comment]
Please plot x vs. time and y vs. time in the imitation learning example with LASA.

File: bolero/environment/optimum_trajectory.py [hunk 1]
```diff
@@ -117,13 +130,14 @@ def get_outputs(self, values):
...
-                self.n_task_dims)
-            values[-self.n_task_dims:] = np.zeros(self.n_task_dims)
+            values[self.n_task_dims:2*self.n_task_dims] = np.zeros(self.n_task_dims)
+            if not self.calcAcc:
+                values[-self.n_task_dims:] = np.zeros(self.n_task_dims)
         else:
             values[:self.n_task_dims] = self.X[self.t - 1]
-            values[self.n_task_dims:-self.n_task_dims] = self.Xd[self.t - 1]
-            values[-self.n_task_dims:] = self.Xdd[self.t - 1]
+            values[self.n_task_dims:2*self.n_task_dims] = self.Xd[self.t - 1]
```
Comment (ChristophPetzoldt) [inline bolero/environment/optimum_trajectory.py]: Keep the use of whitespaces consistent, see PEP8:

> However, in a slice the colon acts like a binary operator, and should have equal amounts on either side (treating it as the operator with the lowest priority). In an extended slice, both colons must have the same amount of spacing applied. Exception: when a slice parameter is omitted, the space is omitted.

> Yes:
> 
> ham[1:9], ham[1:9:3], ham[:9:3], ham[1::3], ham[1:9:]
> ham[lower:upper], ham[lower:upper:], ham[lower::step]
> ham[lower+offset : upper+offset]
> ham[: upper_fn(x) : step_fn(x)], ham[:: step_fn(x)]
> ham[lower + offset : upper + offset]
> 
> No:
> 
> ham[lower + offset:upper + offset]
> ham[1: 9], ham[1 :9], ham[1:9 :3]
> ham[lower : : upper]
> ham[ : upper]

E.g.: 
`values[self.n_task_dims : 2*self.n_task_dims] = self.Xd[self.t-1] `

File: bolero/environment/optimum_trajectory.py [hunk 2]
```diff
@@ -117,13 +130,14 @@ def get_outputs(self, values):
...
+            if not self.calcAcc:
+                values[-self.n_task_dims:] = self.Xdd[self.t - 1]
```
Comment (ChristophPetzoldt) [inline bolero/environment/optimum_trajectory.py]: Would again suggest: `self.Xdd[self.t-1]`

File: bolero/environment/optimum_trajectory.py [hunk 3]
```diff
@@ -135,8 +149,12 @@ def set_inputs(self, values):
...
-        self.Xdd[self.t, :] = values[-self.n_task_dims:]
+        oldVelocity = self.Xd[self.t, :] - values[self.n_task_dims:2*self.n_task_dims]
+        self.Xd[self.t, :] = values[self.n_task_dims:2*self.n_task_dims]
+        if self.calcAcc:
+            self.Xdd[self.t, :] = oldVelocity
+        else:
+            self.Xdd[self.t, :] = values[-self.n_task_dims:]
 
     def step_action(self):
         """Execute step perfectly."""
```
Comment (ChristophPetzoldt) [inline bolero/environment/optimum_trajectory.py]: perfectly?

File: bolero/environment/optimum_trajectory.py [hunk 4]
```diff
@@ -199,6 +233,7 @@ def get_acceleration(self):
         acceleration : array-like, shape (n_steps,)
             the total acceleration (scalar) at all previous timestamps
         """
+
```
Comment (ChristophPetzoldt) [inline bolero/environment/optimum_trajectory.py]: Remove whitespace again.

File: bolero/representation/promp_behavior.py [hunk 5]
```diff
@@ -0,0 +1,497 @@
...
+    yaml.dump(model, model_content)
+    with open(filename, "w") as f:
+        f.write("---\n")
+        f.write(model_content.getvalue())
+        f.write("...\n")
+    model_content.close()
+
+
+class ProMPBehavior(BlackBoxBehavior):
+    """Dynamical Movement Primitive.
```
Comment (ChristophPetzoldt) [inline bolero/representation/promp_behavior.py]: DMP? Change to Probabilistic Movement Primitive.

File: bolero/representation/promp_behavior.py [hunk 6]
```diff
@@ -0,0 +1,497 @@
...
+    dt : float, optional (default: 0.01)
+        Time between successive steps in seconds.
+
+    n_features : int, optional (default: 50)
+        Number of RBF features for each dimension of the promp.
+
+    configuration_file : string, optional (default: None)
+        Name of a configuration file that should be used to initialize the promp.
+        If it is set all other arguments will be ignored.
+    """
```
Comment (ChristophPetzoldt) [inline bolero/representation/promp_behavior.py]: Documentation of `learn_cov` and `use_cov` missing.

File: bolero/representation/promp_behavior.py [hunk 7]
```diff
@@ -0,0 +1,497 @@
...
+    def __init__(self, execution_time=1.0, dt=0.01, n_features=50,
+                 configuration_file=None, learnCovariance = False, useCovar= False):
```
Comment (ChristophPetzoldt) [inline bolero/representation/promp_behavior.py]: Rename to `learn_cov` and `use_cov`

File: bolero/representation/promp_behavior.py [hunk 8]
```diff
@@ -0,0 +1,497 @@
...
+        self.learnCovariance = learnCovariance
+        self.useCovar = useCovar
```
Comment (ChristophPetzoldt) [inline bolero/representation/promp_behavior.py]: Same here.

File: bolero/representation/promp_behavior.py [hunk 9]
```diff
@@ -0,0 +1,497 @@
...
+        ----------
+        n_inputs : int
+            number of inputs
+
+        n_outputs : int
+            number of outputs
+        """
+        #if n_inputs != n_outputs:
+        #    raise ValueError("Input and output dimensions must match, got "
+        #                    "%d inputs and %d outputs" % (n_inputs, n_outputs))
```
Comment (ChristophPetzoldt) [inline bolero/representation/promp_behavior.py]: Why?

File: bolero/representation/promp_behavior.py [hunk 10]
```diff
@@ -0,0 +1,497 @@
...
+                raise ValueError(
+                    "Meta parameter '%s' is not allowed, use one of %r"
+                    % (key, PERMITTED_promp_METAPARAMETERS))
+            setattr(self, key, meta_parameter)
+            
+            if key == "x0":
+                for i in range(len(meta_parameter)):
+                    
+                    conditionPoints += [[0,i,0,meta_parameter[i],0.0001]]
+
```
Comment (ChristophPetzoldt) [inline bolero/representation/promp_behavior.py]: Remove the blank lines here...

File: bolero/representation/promp_behavior.py [hunk 11]
```diff
@@ -0,0 +1,497 @@
...
+        we take the first position as x0.
+
+        Parameters
+        ----------
+        inputs : array-like, shape = (2 * n_task_dims,)
+            Contains positions and velocities in that order.
+            Each type is stored contiguously, i.e. for n_task_dims=2 the order
+            would be: xxvv (x: position, v: velocity).
+        """
+        # just open loop by now
```
Comment (ChristophPetzoldt) [inline bolero/representation/promp_behavior.py]: Add at least a TODO...

File: bolero/representation/promp_behavior.py [hunk 12]
```diff
@@ -0,0 +1,497 @@
...
+        params : array-like, shape = (n_params,)
+            Current weights
+        """
+        random_variables = self.data.mean_
+        if self.learnCovariance:
+            cov = np.array(self.data.covariance_).reshape(len(random_variables),len(random_variables))
+            Dinv =np.linalg.inv(np.sqrt(np.diag(np.diag(cov))))
+            cor = Dinv.dot(cov.dot(Dinv))
+
+            return random_variables + np.sqrt(np.diag(cov)).tolist() + np.arctanh(cor[np.tril_indices(len(random_variables),k=-1)]).tolist()
```
Comment (ChristophPetzoldt) [inline bolero/representation/promp_behavior.py]: Hardly readable and line too long.

File: bolero/representation/promp_behavior.py [hunk 13]
```diff
@@ -0,0 +1,497 @@
...
+        """Reset promp."""
+        if self.x0 is None:
+            self.last_y = np.zeros(self.n_task_dims)
+        else:
+            self.last_y = np.copy(self.x0)
+        self.last_yd = np.copy(self.x0d)
+        self.last_ydd = np.copy(self.x0dd)
+
+        #self.y = np.empty(self.n_task_dims)
+        #self.yd = np.empty(self.n_task_dims)
```
Comment (ChristophPetzoldt) [inline bolero/representation/promp_behavior.py]: Why not reset y, yd and ydd?

File: bolero/representation/promp_behavior.py [hunk 14]
```diff
@@ -0,0 +1,497 @@
...
+        old_means = self.data.mean_[:]
+        self.data.mean_ = params[:len(self.data.mean_)]
+        if self.learnCovariance:
+            D = np.diag(np.square(np.array(params[len(self.data.mean_):2*len(self.data.mean_)])))
+            cor = np.identity(len(self.data.mean_))
+            cor[np.tril_indices(len(self.data.mean_),k=-1)] = np.tanh(params[2*len(self.data.mean_):])
+            cor[np.triu_indices(len(self.data.mean_),k=1)] = np.tanh(params[2*len(self.data.mean_):][::-1])
+            self.data.covariance_ = D.dot(cor.dot(D)).flatten().tolist()
+        self.y, self.yd, self.covars = self.trajectory()
+
```
Comment (ChristophPetzoldt) [inline bolero/representation/promp_behavior.py]: Too many blank lines.

File: bolero/representation/promp_behavior.py [hunk 15]
```diff
@@ -0,0 +1,497 @@
...
+        self.last_ydd = np.copy(self.x0dd)
+
+        #self.y = np.empty(self.n_task_dims)
+        #self.yd = np.empty(self.n_task_dims)
+
+
+        self.last_t = 0.0
+        self.t = 0.0
+
+    def imitate(self, X, Xd=None, Xdd=None, alpha=0.0,
```
Comment (ChristophPetzoldt) [inline bolero/representation/promp_behavior.py]: `alpha` is not used. However, @AlexanderFabisch maybe keep it for consistency with `dmp_behavior.py`?

File: bolero/representation/promp_behavior.py [hunk 16]
```diff
@@ -0,0 +1,497 @@
...
+        The function can be used for debugging purposes.
+
+        Returns
+        -------
+        X : array, shape (n_steps, n_task_dims)
+            Positions
+
+        Xd : array, shape (n_steps, n_task_dims)
+            Velocities
+
```
Comment (ChristophPetzoldt) [inline bolero/representation/promp_behavior.py]: What about Xdd?

File: bolero/representation/promp_behavior.py [hunk 17]
```diff
@@ -0,0 +1,497 @@
...
+        
+        self.data.get_values(x,means,covars)
+        
+        means = means.reshape((self.n_task_dims*2,-1)).transpose(1,0)
+        covars = covars.reshape((-1,self.n_task_dims*2,self.n_task_dims*2))
+        
+        Y = means[:,::2]
+        Yd = means[:,1::2]
+        
+
```
Comment (ChristophPetzoldt) [inline bolero/representation/promp_behavior.py]: Blank lines.

File: bolero/representation/promp_behavior.py [hunk 18]
```diff
@@ -0,0 +1,497 @@
...
+        self.data.get_values(x,means,covars)
+        means = means.reshape((self.n_inputs,-1)).transpose(1,0)
+        covars = covars.reshape((-1,self.n_inputs,self.n_inputs))
+               
+        
+        return means,covars
+
+    
+
+    save = save_promp_model
```
Comment (ChristophPetzoldt) [inline bolero/representation/promp_behavior.py]: What is this?

File: bolero/representation/promp_behavior.py [hunk 19]
```diff
@@ -0,0 +1,497 @@
...
+            vals, vecs = np.linalg.eigh(cov)
+            order = vals.argsort()[::-1]
+            return vals[order], vecs[:,order]
+        nstd = 2
+        cov = np.empty((2,2))
+        for k in range(0,len(means)):            
+            cov[0,0] = covariances[k,0,0]
+            cov[0,1] = covariances[k,0,2]
+            cov[1,0] = covariances[k,2,0]
+            cov[1,1] = covariances[k,2,2]
```
Comment (ChristophPetzoldt) [inline bolero/representation/promp_behavior.py]: PEP8: whitespace

File: examples/representation/lasa/plot_promp_lasa.py [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+    return (x,y)
+
+def learn(x,y):
+    traj = ProMPBehavior(1.0, 1.0/999.0, numWeights,learnCovariance=True,useCovar=True)
+    traj.init(4,4)
+    traj.imitate(y.transpose(2,1,0))  
+    return traj
+
+def draw(x,y,traj,useCovariance,useSamples):
+	ax = plt.subplot(121)
```
Comment (ChristophPetzoldt) [inline examples/representation/lasa/plot_promp_lasa.py]: useSamples not used

File: examples/representation/lasa/plot_promp_lasa.py [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+	useSamples = bool(int(argv[2]))
+	x,y = load(int(argv[0]))
+	traj = learn(x,y)
+	draw(x,y,traj,useCovariance,useSamples)
+	plt.show()
+
+if __name__ == "__main__":
+    main(sys.argv[1:])
+
+
```
Comment (ChristophPetzoldt) [inline examples/representation/lasa/plot_promp_lasa.py]: In this file in general: inconsistency with tabs and whitespace. Please use whitespace.

File: bolero/representation/promp_behavior.py (hunk 12 above)
Comment (AlexanderFabisch) [reply bolero/representation/promp_behavior.py]: also: why does it have to be a list? you can concatenate numpy arrays

File: bolero/representation/promp_behavior.py (hunk 18 above)
Comment (AlexanderFabisch) [reply bolero/representation/promp_behavior.py]: that's ok: the member function save is the same as save_promp_model

File: manifest.xml [hunk 22]
```diff
@@ -32,4 +32,6 @@
   <rosdep name="python-matplotlib" />
   <rosdep name="python-sklearn" />
   <depend package="learning/bolero/bolero/wrapper" />
+  <depend package="learning/bolero/src/representation/promp" />
```
[the remark this answers is not in the document]
Comment (AlexanderFabisch) [reply manifest.xml]: TODO for me: add to buildconf after merge

File: bolero/datasets/lasa.py [hunk 23]
```diff
@@ -0,0 +1,91 @@
...
+
+    shape_name : string
+        Name of the Matlab file from which we load the demonstrations
+        (without suffix)
+    """
+
+    dataset_path = os.path.expanduser("~")
+    dataset_path += os.sep + "bolero_data" + os.sep
+
+    if not os.path.isdir(dataset_path):
```
Comment (AlexanderFabisch) [inline bolero/datasets/lasa.py]: could you put that in a separate function?

File: bolero/environment/optimum_trajectory.py (hunk 3 above)
Comment (AlexanderFabisch) [reply bolero/environment/optimum_trajectory.py]: without noise

**Comment by ChristophPetzoldt** [comment]
@bpoppinga Could you please add unit tests to a file called `bolero/representation/test/test_promp_behavior.py`?
Similar to `test_dmp_behavior.py` or `test_dmp_seq.py, respectively.

**Comment by ChristophPetzoldt** [comment]
Apart from that, the installation now runs successfully and I can run the example scripts.

File: bolero/datasets/lasa.py [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+    X = np.transpose([P for P, _, _ in tmp], [1, 2, 0])
+    Xd = np.transpose([V for _, V, _ in tmp], [1, 2, 0])
+    Xdd = np.transpose([A for _, _, A in tmp], [1, 2, 0])
+
+    dt = float(demos[0][0, 0][4])
+
+    return X, Xd, Xdd, dt
+
+
+def get_common_dataset_path():
```
Comment (ChristophPetzoldt) [inline bolero/datasets/lasa.py]: Stores the bolero_data folder directly in the home folder, when I execute `plot_lasa_promp.py`.
Is this the desired behavior? I would suggest to store into the bolero main folder.

Comment (AlexanderFabisch) [reply bolero/datasets/lasa.py]: No, it should be `.bolero_data` - a hidden file in the home folder.

Comment (bpoppinga) [reply bolero/datasets/lasa.py]: I think that Scikit-Learn uses a non hidden folder: 
if data_home is None:
        data_home = environ.get('SCIKIT_LEARN_DATA',
                                join('~', 'scikit_learn_data'))
    data_home = expanduser(data_home)
    if not exists(data_home):
        makedirs(data_home)
return data_home
https://github.com/scikit-learn/scikit-learn/blob/f0ab589f/sklearn/datasets/base.py#L325

Comment (AlexanderFabisch) [reply bolero/datasets/lasa.py]: OK, never mind, I think it is OK

**Comment by AlexanderFabisch** [comment]
Merged with 151ed :beers:

**Pull request closed without merging**